### PR TITLE
Fix some unstable tests

### PR DESCRIPTION
**Pull request by erszcz**
This PR fixes some tests which spuriously fail Travis builds. ~I currently have a list at least this long~ The list below contains links to all failures from builds which finally passed for PRs #1662 and #1846:

- [ ] `bosh_SUITE`: [`interleave_requests_statem`][b1]
- [ ] `connect_SUITE:'clients_can_connect_with_DHE-RSA-AES256-SHA'/1`
- `ejabberdctl_SUITE`:
  * [ ] [`stats_global`][e1]
  * [x] `num_active_users`
- [x] `login_SUITE`: [1][l1], [2][l2]
- [x] `metrics_register_SUITE`: [`unregister`][mr1]
- [ ] `mod_global_distrib_SUITE`: [`test_pm_with_graceful_reconnection_to_different_server`][mgd1], [`test_muc_conversation_history`][mgd2]
- [ ] `muc_http_api_SUITE`: [1][mha1], [2][mha2], [3][mha3], [4][mha4], [5][mha5]
- [ ] `push_integration_SUITE`: [1][pi1]
- [x] `sm_SUITE`: [`resume_session_state_send_message 1`][sm1], [`resume_session_state_send_message 2`][sm2], [3][sm3]

[b1]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/ldap_mnesia.19.3/big/ct_run.test%40travis-job-d9da47b3-4e7a-4685-a3a9-4f15f8e53dcc.2018-05-21_15.38.40/big_tests.tests.bosh_SUITE.logs/run.2018-05-21_15.39.27/bosh_suite.interleave_requests_statem.718.html

[e1]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/riak_mnesia.18.3/big/ct_run.test%40travis-job-e34d5ce9-4958-48de-81b9-8614c6e7d29e.2018-05-21_20.20.25/big_tests.tests.ejabberdctl_SUITE.logs/run.2018-05-21_20.28.27/suite.log.html

[e3]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/riak_mnesia.18.3/big/ct_run.test%40travis-job-c5404850-29b9-4c51-9566-d2a7308ed3bc.2018-05-21_15.54.23/big_tests.tests.ejabberdctl_SUITE.logs/run.2018-05-21_16.02.22/suite.log.html

[e4]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1846/4603/riak_mnesia.18.3/big/ct_run.test%40travis-job-c1e809ec-ac16-4044-8111-bc9ebeeca37e.2018-05-14_13.46.30/big_tests.tests.ejabberdctl_SUITE.logs/run.2018-05-14_13.54.44/suite.log.html

[e5]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1846/4603/riak_mnesia.18.3/big/ct_run.test%40travis-job-a4096f71-26a8-4498-b119-11343a203e49.2018-05-14_12.53.27/big_tests.tests.ejabberdctl_SUITE.logs/run.2018-05-14_13.01.34/suite.log.html

[e6]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1846/4603/riak_mnesia.18.3/big/ct_run.test%40travis-job-9e2f0ab0-cac8-4375-9f91-667d42f725ed.2018-05-14_11.17.49/big_tests.tests.ejabberdctl_SUITE.logs/run.2018-05-14_11.25.57/suite.log.html

[l1]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1846/4603/riak_mnesia.18.3/big/ct_run.test%40travis-job-a4096f71-26a8-4498-b119-11343a203e49.2018-05-14_12.53.27/big_tests.tests.login_SUITE.logs/run.2018-05-14_13.02.12/suite.log.html

[l2]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1846/4603/riak_mnesia.18.3/big/ct_run.test%40travis-job-9e2f0ab0-cac8-4375-9f91-667d42f725ed.2018-05-14_11.17.49/big_tests.tests.login_SUITE.logs/run.2018-05-14_11.26.35/suite.log.html

[mgd1]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/mysql_redis.19.3/big/ct_run.test%40travis-job-b859f252-ed42-4399-8fad-b13391ae447b.2018-05-21_16.07.18/big_tests.tests.mod_global_distrib_SUITE.logs/run.2018-05-21_16.24.35/suite.log.html

[mgd2]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/odbc_mssql_mnesia.19.3/big/ct_run.test%40travis-job-9eadd38a-d144-40de-ae8f-38c12650da4a.2018-05-21_15.08.39/big_tests.tests.mod_global_distrib_SUITE.logs/run.2018-05-21_15.28.27/suite.log.html

[mha1]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/riak_mnesia.18.3/big/ct_run.test%40travis-job-ca393cf7-af88-4bad-9d4a-c6f0a8e6ebfd.2018-05-21_16.53.39/big_tests.tests.muc_http_api_SUITE.logs/run.2018-05-21_17.07.10/suite.log.html

[mha2]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/riak_mnesia.18.3/big/ct_run.test%40travis-job-3a659b22-5376-4a6c-a065-c67b895139c9.2018-05-21_17.53.11/big_tests.tests.muc_http_api_SUITE.logs/run.2018-05-21_18.06.54/suite.log.html

[mha3]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1846/4603/riak_mnesia.18.3/big/ct_run.test%40travis-job-c1e809ec-ac16-4044-8111-bc9ebeeca37e.2018-05-14_13.46.30/big_tests.tests.muc_http_api_SUITE.logs/run.2018-05-14_14.00.50/suite.log.html

[mha4]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1846/4603/riak_mnesia.18.3/big/ct_run.test%40travis-job-a982e6eb-06e4-4ab2-9b74-904f41183c98.2018-05-15_09.07.33/big_tests.tests.muc_http_api_SUITE.logs/run.2018-05-15_09.20.47/suite.log.html

[mha5]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1846/4603/riak_mnesia.18.3/big/ct_run.test%40travis-job-634ce4e4-1522-4bd8-a742-6db46d35e838.2018-05-14_12.07.40/big_tests.tests.muc_http_api_SUITE.logs/run.2018-05-14_12.20.57/suite.log.html

[mr1]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/riak_mnesia.18.3/big/ct_run.test%40travis-job-ca393cf7-af88-4bad-9d4a-c6f0a8e6ebfd.2018-05-21_16.53.39/big_tests.tests.metrics_register_SUITE.logs/run.2018-05-21_17.04.56/suite.log.html

[pi1]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/riak_mnesia.18.3/big/ct_run.test%40travis-job-ca393cf7-af88-4bad-9d4a-c6f0a8e6ebfd.2018-05-21_16.53.39/big_tests.tests.push_integration_SUITE.logs/run.2018-05-21_17.08.33/suite.log.html

[sm1]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/mysql_redis.19.3/big/ct_run.test%40travis-job-f88b73b6-ef4b-47f1-9d22-5de894bab919.2018-05-21_15.07.02/big_tests.tests.sm_SUITE.logs/run.2018-05-21_15.22.45/suite.log.html

[sm2]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1662/4692/mysql_redis.19.3/big/ct_run.test%40travis-job-b859f252-ed42-4399-8fad-b13391ae447b.2018-05-21_16.07.18/big_tests.tests.sm_SUITE.logs/run.2018-05-21_16.23.43/suite.log.html

[sm3]: http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1846/4603/riak_mnesia.18.3/big/ct_run.test%40travis-job-9e2f0ab0-cac8-4375-9f91-667d42f725ed.2018-05-14_11.17.49/big_tests.tests.sm_SUITE.logs/run.2018-05-14_11.35.00/suite.log.html

### CODE REVIEW

**Comment by erszcz** [comment]
Main offender in `ejabberdctl_SUITE` is `stats_global` - 5 fails out of 5 linked builds.

**Comment by arcusfelis** [comment]
Some broken tests are reported and have a tag [ci-errors](https://github.com/esl/MongooseIM/issues?q=is%3Aissue+is%3Aopen+label%3Aci-errors).

**Comment by erszcz** [comment]
`muc_http_api_SUITE` - not a single failure in 500 test repetitions run in each of the presets... [build 4732](https://travis-ci.org/esl/MongooseIM/jobs/382684628)

**Comment by erszcz** [comment]
Fantastic, [build 4754](https://travis-ci.org/esl/MongooseIM/builds/383108814) testing just `ejabberdctl_SUITE` is also all green... The strongest offenders, when put under closer inspection, are not causing any trouble.

**Comment by erszcz** [comment]
I'm rebasing this branch - the above comments link to builds performed to troubleshoot the spurious failures. After rebase, the commits themselves won't be listed as part of this PR.

**Comment by arcusfelis** [comment]
Another resume_session_state_send_message http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1879/4755/riak_mnesia.18.3

**Comment by erszcz** [comment]
[Job 4760.2](https://travis-ci.org/esl/MongooseIM/jobs/383266086) run from #1888 tests a promising approach - we can make CT automatically retry a whole test group if any of the tests fails. As shown in [4760.2 results for group `odbc_async_pool_prefs_cases`](http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1888/4760/mysql_redis.19.3/big/ct_run.test@travis-job-36c72021-1d6a-4687-95a5-919c672f41a5.2018-05-24_17.50.24/big_tests.tests.mam_SUITE.logs/run.2018-05-24_17.54.33/suite.log.html) - tests 646 to 659 - the group was restarted and the second run of the group's tests succeeded.

The missing bit is our Travis script logic determining whether running big tests failed or not - we have a non-zero number of failed test cases, although ultimately the group and therefore the whole run succeeded.

**Comment by erszcz** [comment]
For the record: https://www.strawpoll.me/15765236/r

**Comment by ludwikbukowski** [comment]
Regarding:
- [ ] `push_integration_SUITE`: [1][pi1]

I experienced the same few times on my local machine and it was due to bugs from FCM mock:
https://github.com/rslota/fcm-http2-mock-server
```
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/events.py", line 125, in _run
    self._callback(*self._args)
  File "/usr/lib/python3.5/asyncio/selector_events.py", line 669, in _read_ready
    self._protocol.data_received(data)
  File "/usr/lib/python3.5/asyncio/sslproto.py", line 506, in data_received
    self._app_protocol.data_received(chunk)
  File "/server/asyncio-http2-wsgi-server.py", line 156, in data_received
    events = self.conn.receive_data(data)
  File "/usr/local/lib/python3.5/dist-packages/h2/connection.py", line 1522, in receive_data
    events.extend(self._receive_frame(frame))
  File "/usr/local/lib/python3.5/dist-packages/h2/connection.py", line 1545, in _receive_frame
    frames, events = self._frame_dispatch_table[frame.__class__](frame)
  File "/usr/local/lib/python3.5/dist-packages/h2/connection.py", line 1788, in _receive_ping_frame
    ConnectionInputs.RECV_PING
  File "/usr/local/lib/python3.5/dist-packages/h2/connection.py", line 238, in process_input
    "Invalid input %s in state %s" % (input_, old_state)
h2.exceptions.ProtocolError: Invalid input ConnectionInputs.RECV_PING in state ConnectionState.CLOSED
Exception in callback _SelectorSocketTransport._read_ready()
handle: <Handle _SelectorSocketTransport._read_ready()>
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/h2/connection.py", line 2038, in _decode_headers
    return decoder.decode(encoded_header_block, raw=True)
  File "/usr/local/lib/python3.5/dist-packages/hpack/hpack.py", line 444, in decode
    data_mem[current_index:]
  File "/usr/local/lib/python3.5/dist-packages/hpack/hpack.py", line 492, in _decode_indexed
    header = HeaderTuple(*self.header_table.get_by_index(index))
  File "/usr/local/lib/python3.5/dist-packages/hpack/table.py", line 130, in get_by_index
    raise InvalidTableIndex("Invalid table index %d" % index)
hpack.exceptions.InvalidTableIndex: Invalid table index 22
```

**Comment by erszcz** [comment]
The `ct_groups_summary_hook` writes a `group.summary` file similar to the `suite.summary` generated by Common Test. [Here's a listing with the example file for the failed `job 4763.3 mod_blocking_SUITE`](http://opensource.erlang-solutions.com/mongooseim-ct-reports/s3_reports.html?prefix=PR/1874/4763/odbc_mssql_mnesia.19.3/big/ct_run.test@travis-job-ae09e926-fc35-49a0-a213-fd04a8ea0681.2018-05-25_16.09.08/big_tests.tests.mod_blocking_SUITE.logs/run.2018-05-25_16.19.24/). This means we have a failed test, but no failed groups (since the group in which a test failed was restarted and the consecutive run succeeded).

To make this approach to restarting failed tests sensible we should do two more things:
- [ ] extend https://github.com/esl/MongooseIM/blob/master/tools/travis-test.sh to interpret `groups.summary` for marking a job as green or red (the above example job is red, though it would be green if we checked `groups.summary`)
- [ ] either list failed cases in a corresponding PR (ideal) or at the very least print out from `travis-test.sh` (a poor man's solution)

File: big_tests/tests/amp_big_SUITE.erl [hunk 1]
```diff
@@ -28,26 +28,28 @@ enabled_group_names() ->
...
-     {offline_success, [], [{group, G} || G <- subgroup_names()]},
-     {offline_failure, [], [{group, G} || G <- subgroup_names()]}
-    ] ++
-        [{G, [parallel, shuffle], notify_deliver_test_cases()}
-         || G <- notify_deliver_group_names()] ++
-        [{G, [parallel, shuffle], error_deliver_test_cases()}
-         || G <- error_deliver_group_names()] ++
-        [{G, [parallel, shuffle], drop_deliver_test_cases()}
-         || G <- drop_deliver_group_names()].
+    Gs = ([{basic, [parallel], [{group, G} || G <- subgroup_names()] ++ basic_test_cases()},
```
Comment (arcusfelis) [inline big_tests/tests/amp_big_SUITE.erl]: Putting the list into separate function makes it's easier to read.

```erlang
groups() ->
    ct_helper:repeat_all_until_all_ok(groups_list()).

groups_list() ->
   [{basic, [parallel], [{group, G} || G <- subgroup_names()] ++ basic_test_cases()},
   ....
```

Comment (arcusfelis) [reply big_tests/tests/amp_big_SUITE.erl]: And we don't need to reindent.

**Comment by arcusfelis** [comment]
Side note: sometimes if we start/stop modules and/or start/stop the whole application, these actions can fail for some reason. But we don't handle the errors.

Basically, stop/start mod_mam can sometimes fail. For example, because of code_purger doing something strange in mod_muc:init/1

**Comment by ludwikbukowski** [comment]
for `login_SUITE` here is PR https://github.com/esl/MongooseIM/pull/1914
waiting for travis

**Comment by mongoose-im** [comment]
[4934.1](https://travis-ci.org/esl/MongooseIM/jobs/393510666) / Erlang 19.3 / internal_mnesia / a7bf2dd59d3e9045fee28b79855379bb1067101c
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1874/4934/internal_mnesia.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1874/4934/internal_mnesia.19.3/big/ct_run.test%40travis-job-9f28da5a-0381-4470-9cff-2feebc3e6f92.2018-06-18_08.02.34/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1874/4934/internal_mnesia.19.3/small/ct_run.mongooseim@localhost.2018-06-18_07.55.38/lib.mongooseim.logs/run.2018-06-18_07.55.39/suite.log.html)
**OK: 1059** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[4944.5](https://travis-ci.org/esl/MongooseIM/jobs/394504413) / Erlang 19.3 / elasticsearch_and_cassandra_mnesia / e44d2689b151a950bfcff939d5a2273a9899210b
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1874/4944/elasticsearch_and_cassandra_mnesia.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1874/4944/elasticsearch_and_cassandra_mnesia.19.3/big/ct_run.test%40travis-job-a8472051-1634-40a3-98ec-e9571e12ea0a.2018-06-20_11.17.31/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1874/4944/elasticsearch_and_cassandra_mnesia.19.3/small/ct_run.mongooseim@localhost.2018-06-20_11.10.24/lib.mongooseim.logs/run.2018-06-20_11.10.24/suite.log.html)
**OK: 445** / Failed: 0 / **User-skipped: 8** / Auto-skipped: 0

**Pull request merged**